File: [0421][Medium][Maximum_XOR_of_Two_Numbers_in_an_Array]/Maximum_XOR_of_Two_Numbers_in_an_Array_4.py
#-*- coding: utf-8 -*-  
# leetcode time     cost : 120 ms
# leetcode memory   cost : 13.7 MB 
'''
General form of solution is to find two numbers which have the highest bits complementing each other (0 and 1). 
Then store that max, look at the next less significant bit and see if another two numbers complement each other 
(which includes the bits sets from the previous max). If so, this number will
be greater than the previous maximum. Steps:

Find the new possible maximum by setting the next bit to 1 on the current maximum
use a mask with all 1s up to the current bit i and & all numbers with this mask to see which bit in that number up to bit i are set as 1.
Critical part: to solve this in O(n) it's important to know that If a ^ b = c, 
then a ^ c = b and c ^ b = a. So, if we are looking for a particular number (in our case a possible maximum) we can do the following search: 
iterate in the numbers that we &'d with the mask. If our potental maximum is c, and our current number is a, 
we're looking for another number b that XOR'd with a gives c. Since we also know from above that a ^ c = b, 
we can just look for b it directly in our numbers (i.e. b = a ^ c, or in my code, bit ^ possible_mx).
'''
class Solution:
    def findMaximumXOR(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        mx,mask=0,0
        for i in range(31,-1,-1):
            possible_mx = mx | 1 << i
            mask = mask | 1 << i
            bits=set()
            for num in nums:
                bits.add(num & mask)
            for bit in bits:
                if bit ^ possible_mx in bits:
                    mx = possible_mx
                    break
        return mx 

def main():
    obj = Solution()
    nums = [3,10,5,25,2,8]      # expect is 28
    result = obj.findMaximumXOR(nums)
    print(result)
    
if __name__ =='__main__':
    main()  

File: [0421][Medium][Maximum_XOR_of_Two_Numbers_in_an_Array]/Maximum_XOR_of_Two_Numbers_in_an_Array_4.java
// leetcode time     cost : 10 ms
// leetcode memory   cost : 51.6 MB 
class Solution {
    class TrieNode {
        int val;
        TrieNode zero, one;
        boolean isEnd;
    }
    class TrieTree {
        TrieNode root;
        public TrieTree() {
            root = new TrieNode();
        }
        public void insert(int num) {
            TrieNode cur = root;
            int j = 1 << 30;
            for (int i = 0; i < 31; i++) {
                // 根据num在j位置的数目判断应该向0还是向1
                int b = (j & num) == 0 ? 0 : 1;
                if (b == 0 && cur.zero == null) {
                    cur.zero = new TrieNode();
                }
                if (b == 1 && cur.one == null) {
                    cur.one = new TrieNode();
                }
                cur = b == 0 ? cur.zero : cur.one;
                j >>= 1;
            }
            cur.isEnd = true;
            cur.val = num;
        }
    }

    public int findMaximumXOR(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return 0;
        }
        TrieTree tree = new TrieTree();
        for (int n : nums) {
            tree.insert(n);
        }
        // 获取真正需要开始判断的root
        TrieNode cur = tree.root;
        while (cur.one == null || cur.zero == null) {
            cur = cur.zero != null ? cur.zero : cur.one;
        }
        return maxHelper(cur.one, cur.zero);

    }

    /**
     * 分治法，原则就是尽量使两个分支的高位不同
     * one是1分支，zero是0分支，可以看做图中的左区和右区
     * 1. 当1分支的下一位只有1时，找0分支的0，若没有，就找0分支的1
     * 2. 当1分支的下一位只有0时，找0分支的1，若没有，就找0分支的0
     * 3. 当1分支的下一位0，1均有时，看0分支：如果0分支只有1，则1分支走0，反之则走1
     * 4. 当0，1两个分支均有两个下一位时，尝试【1分支走1，0分支走0】和【1分支走0，0分支走1】两条路线并取最大值
     * */
    private int maxHelper(TrieNode one, TrieNode zero) {
        if (one.isEnd && zero.isEnd) {
            return one.val ^ zero.val;
        }
        if (one.zero == null) {
            return maxHelper(one.one, zero.zero == null ? zero.one : zero.zero);
        } else if (one.one == null) {
            return maxHelper(one.zero, zero.one == null ? zero.zero : zero.one);
        } else if (zero.zero == null) {
            return maxHelper(one.zero, zero.one);
        } else if (zero.one == null) {
            return maxHelper(one.one, zero.zero);
        } else {
            return Math.max(maxHelper(one.one, zero.zero), maxHelper(one.zero, zero.one));
        }
    }
}